File: Assignment-2/hashing.py
"""
Program to implement hash table in python.

Authors:
 Ujjawal Aggarwal - 40183962
 Kshitij Yerande - 40194579
 
"""
from timeit import default_timer as timer
import tracemalloc
import sys



"""
Node class to store data of each key

Attribute:
    Item : key value to be stored.
    next : reference to next item
"""
class Node:
    def __init__(self,item):
        self.item = item
        self.next = None 


"""
HashTable class to store data.

Attribute:
    capacity : size of hash table.
    size : number of elements stored in hash table.
    buckets: number of hashing buckets available.

Methods:
    hash : computes hash value fo given key
    put : adds key to hash table
    contains: checks whether key is present in hash table.
    populate: populates hash table from file
    print: prints hash table on console.

"""
class HashTable:
    def __init__(self,initial_capacity):
        self.capacity = initial_capacity
        self.size = 0
        self.buckets = [None] * self.capacity
        
    def hash(self, key):
        hashval = key % self.capacity
        return hashval
    
    def put(self, item):
            
        index = self.hash(item)
        
        node = self.buckets[index]
        
        if node is None:
            self.buckets[index] = Node(item)
            self.size+=1
            return
        
        if self.contains(item)[0]:
            return
            
        newnode = Node(item)
        newnode.next = node
        self.buckets[index] = newnode
        
        self.size+=1
    
    def contains(self,item):
        counter = 0
        index = self.hash(item)
        
        node = self.buckets[index]
        
        if node is None:
            counter += 1
            return [False,counter]
        
        prev = node
        
        if node.item == item:
                counter += 1 
                return [True,counter]
        else: 
            while node is not None:
                prev = node
                node = node.next
                counter += 1
                
                if prev.item == item:
                    return [True,counter]
        return [False,counter]
    
    def populate(self,filename):
        with open(filename) as f:
            lines = f.readlines()
            for item in lines:
                self.put(int(item))
    
    def print(self):
        k=0
        for i in self.buckets:
            print("Bucket:",(k+1))
            while i is not None:
                print(str(i.item)+" ")
                i = i.next
            print("\n")
            k+=1


"""
Method : readFile - Method to read data from file to be searched.

Parameters :
    filename - name of the data file to be searched

Returns:
    arr - list of numbers to be searched

"""
def readFile(filename):

    arr=[]
    
    with open(filename) as f:
        arr = f.readlines()
        arr = [int(i) for i in arr]
    
    return arr


"""
Method : writeFile - Method to write data to file
                     Creats a file in current directory.

Parameters :
    arr - string data
"""
def writeFile(arr):
    filename = "output_hashing_python.txt"
    with open(filename,"a") as f:
        f.write(str(arr)+'\n')


"""
Main entry of the program.

Arguments:
    file name - data file to be hashed.
"""
if __name__ == "__main__":

    if len(sys.argv) < 2:
        raise TypeError("Arguments missing.")
    
    filename = sys.argv[1]
    
    hashtable = HashTable(600011)
    tracemalloc.start()
    start_populate_time = timer()
    hashtable.populate(filename)
    end_populate_time = timer()
    current, peak = tracemalloc.get_traced_memory()
    tracemalloc.stop()
    
    print("Hash Table created.")
    
    memory_usage=peak / 10**6
    
    writeFile(str("Statistics Hash Table in Python\n"))
    
    writeFile(str("HashTable size:"+str(hashtable.size)))
    
    populate_time = (end_populate_time - start_populate_time) * 1000
    writeFile(str("Time to populate data in hash table:"+str(populate_time)+" milliseconds"))
    writeFile(str("Memory required to populate data in hash table:"+str(memory_usage)+" MB"))
    
    writeFile(str("\nSearch Statistics\n"))
    
    searcharr = readFile('randSearch.txt')
    
    searchdict = {True:[],False:[]}
    
    for i in searcharr:
        start_s_time = timer()
        out = hashtable.contains(i)
        end_s_time = timer()
        
        search_time = (end_s_time - start_s_time) * 1000
        
        if out[0] == True:
            writeFile(str(str(i) + " founded in "+ str(search_time)+" milliseconds with "+str(out[1])+" number of accesses."))
        else:
            writeFile(str(str(i) + " not founded in "+ str(search_time)+" milliseconds with "+str(out[1])+" number of accesses."))
        
        searchdict[out[0]].append(search_time)
    
    if len(searchdict[True]) == 0:
        avg_success_time = 0
    else:
        avg_success_time = sum(searchdict[True]) / len(searchdict[True])
    
    if len(searchdict[False]) == 0:
        avg_unsuccess_time = 0
    else:
        avg_unsuccess_time = sum(searchdict[False]) / len(searchdict[False])
    
    writeFile(str("\nAverage successfull search time:"+str(avg_success_time)+" milliseconds"))
    writeFile(str("Average un successfull search time:"+str(avg_unsuccess_time)+" milliseconds"))
    
    print("Log file created")

File: Assignment-2/Hashing.java

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;
import java.lang.management.ManagementFactory;
import com.sun.management.OperatingSystemMXBean;

/**
 * Hashing.
 *
 * @author Seung Hyun Hong, Nastaran Naseri
 */
public class Hashing {
    static final Runtime runtime = Runtime.getRuntime();

    /** The Constant MEGABYTE. */
    private static final long MEGABYTE = 1024L * 1024L;

    /** The Constant SECOND. */
    private static final long MILLISECOND = 1000000;


    /**
     * The main method.
     *
     * @param args the arguments
     */
    public static void main(String[] args) {
        // get raw data file name
        String filename = args[0];
        // create Hashing object
        Hashing h = new Hashing();
        // variable for output file
        String outputFile = "output_hashing_java";

        // run garbage collector before calculate memory usage
        runtime.gc();
        //memory used before creation of BST
        long beforeUsedMem = runtime.totalMemory() - runtime.freeMemory();
        double startTimeReading = System.nanoTime();
        // read raw data from the file and put it in array list
        ArrayList<Long> arr = readData(filename);

        double endTimeReading = System.nanoTime();
        double durationReading = (endTimeReading - startTimeReading);

        double startTimeHashing = System.nanoTime();

        HashingTable ht = h.new HashingTable();
        // Put raw data into HashTable
        for (int i = 0; i < arr.size(); i++) {
            ht.put(ht,arr.get(i));
        }
        double endTimeHashing = System.nanoTime();
        double durationHashing = (endTimeHashing - startTimeHashing);

        //memory used after creation of BST
        long afterUsedMem = runtime.totalMemory() - runtime.freeMemory();
        //actual memory used while creating BST
        long memory = afterUsedMem-beforeUsedMem;

        // get search variables from the file into arraylist
        ArrayList<Long> search = readData(args[1]);

        // get search result
        ArrayList<String> searchTime = searchHasing(ht, search);

        // create out put file for statistics
        ArrayList<String> stat = new ArrayList<String>();
        stat.add("Statistics for Hashing with JAVA\n");
        stat.add("Number of elements: " + arr.size());
        stat.add("Reading file execution time: " + durationReading / MILLISECOND + " milliseconds");
        stat.add("Hashing execution time " + durationHashing / MILLISECOND + " milliseconds");
        stat.add("Memory Used for create hash table: " + memory / MEGABYTE + " megabytes\n");

        for (int i = 0; i < searchTime.size(); i++) {
            stat.add(searchTime.get(i));
        }
        writeData(stat, outputFile);
    }

    /**
     * Read raw data from text file.
     *
     * @param name the file name
     * @return the array list
     */
    public static ArrayList<Long> readData(String name) {
        ArrayList<Long> arr = new ArrayList<Long>();
        try {
            Scanner fis = new Scanner(new FileInputStream(name));
            // read each line
            while (fis.hasNextLine()) {
                arr.add(fis.nextLong());
            }
            fis.close();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
        return arr;
    }

    /**
     * Write sorted data into text file.
     *
     * @param <T> the generic type
     * @param arr the sorted data array list
     * @param name the output file name
     */
    public static <T> void writeData(ArrayList<T> arr, String name) {
        try {
            FileWriter myWriter = new FileWriter(name + ".txt");
            for (int i = 0; i < arr.size(); i++) {
                myWriter.write(arr.get(i) + "\n");
            }
            myWriter.close();
            System.out.println("Output file created ");
        } catch (IOException e) {
            System.out.println("error");
            e.printStackTrace();
        }
    }



    /**
     * Search hasing.
     *
     * @param ht the HashTable
     * @param search the search data
     * @return the array list of string that contain result of searches
     */
    public static ArrayList<String> searchHasing(HashingTable ht, ArrayList<Long> search) {
        double tTime = 0;
        int tCount = 0;
        double fTime = 0;
        int fCount = 0;
        ArrayList<String> res = new ArrayList<>();
        res.add("\nSearch in Hash Table\n");
        for (int i = 0; i < search.size(); i++) {
            double start = System.nanoTime();
            Map<Integer, Integer> chk = ht.search(search.get(i));
            double end = System.nanoTime();
            if (chk.containsKey(1)) {
                tTime += (end - start) / MILLISECOND;
                tCount++;
                res.add(search.get(i) + "  founded in " + (end - start) / MILLISECOND + " milliseconds with " + chk.get(1)
                        + " number of accesses.");

            } else {
                fTime += (end - start) / MILLISECOND;
                fCount++;
                res.add(search.get(i) + "  not founded in " + (end - start) / MILLISECOND + " milliseconds with " + chk.get(0)
                        + " number of accesses.");
            }
        }
        res.add("\nAverage time for successful search is " + (tTime / tCount) + " milliseconds");
        res.add("Average time for un-successful search is " + (fTime / fCount)+ " milliseconds");
        res.add("Average time for total search is " + ((tTime / tCount) + (fTime / fCount)) + " milliseconds");

        return res;
    }

    /**
     * The Class HashingTable.
     */
    class HashingTable {

        /** The table size. */
        private int tableSize = 600011;

        /** The hash table. */
        private Node[] hashTable = new Node[tableSize];

        /**
         * The Class Node.
         */
        class Node {

            /** The key. */
            long key;

            /** The next. */
            Node next;
        }

        /**
         * Put key in HashTable
         *
         * @param ht the HeahTable
         * @param key the key
         */
        public void put(HashingTable ht, long key) {
            int location = (int) (Math.abs(key) % tableSize);
            Node newNode = new Node();
            newNode.key = key;
            Map<Integer, Integer> chk = ht.search(key);

            if(chk.containsKey(0)) {
                if (hashTable[location] == null) {
                    hashTable[location] = newNode;
                } else {
                    Node runner = hashTable[location];
                    hashTable[location] = newNode;
                    newNode.next = runner;
                }
            }
        }

        /**
         * Search the key
         *
         * @param key the key
         * @return the map for successful/ un-successful and number of accesses
         */
        public Map<Integer, Integer> search(long key) {
            Map<Integer, Integer> res = new HashMap<>();
            Integer location = (int) (Math.abs(key) % tableSize);
            int cnt = 0;
            Node runner = hashTable[location];

            while (runner != null) {
                cnt++;
                if (runner.key == key) {
                    res.put(1, cnt);
                    return res;
                }
                runner = runner.next;
            }
            res.put(0, cnt);
            return res;
        }
    }
}
